### ISSUE TITLE
Enhancement: autosize to fit a track

### ISSUE THREAD

**Issue by marcoXbresciani** [y20k/trackbook#60]
Once a track is saved, it would be nice to either have an automatic autosize to fit the whole path, or a button to allow such autosize so that the whole tracked path will fit the screen and can be seen easily as a whole.

**Comment by y20k** [comment]
Sounds like a good idea. I will check, how to implement this. 

But beware. The development speed for this app is quite slow :-) I am doing this in my free time, of which I do have too much at the moment (family & job ...).

**Comment by marcoXbresciani** [comment]
> Once a track is saved, it would be nice to either have an automatic autosize to fit the whole path, or a button to allow such autosize so that the whole tracked path will fit the screen and can be seen easily as a whole.

A simple double tap could do the trick as well.

**Comment by marcoXbresciani** [comment]
> Sounds like a good idea. I will check, how to implement this.
> 
> But beware. The development speed for this app is quite slow :-) I am doing this in my free time, of which I do have too much at the moment (family & job ...).

Not sure about the code (I'll try to take a look but I'm not so much into apps... old school Java here), but if you have the list of waypoints, you "simply" have to search for the most western, the most northern, eastern and southern ones and then take their coordinates to fit the track window/zoom.

**Comment by Hob111** [comment]
This is a good idea. I'd prefer a track opened as it does now but could be autosized by a double tap as suggested or maybe a long press would be better.

**Comment by y20k** [comment]
@marcoXbresciani we could split the coding job. I would implement the UI. You could write a function the return the edge coordinates (BoundingBox?) and the center of a recorded [track](https://github.com/y20k/trackbook/blob/master/app/src/main/java/org/y20k/trackbook/core/Track.java).

If I understand the API correct. One would need to do two things in the UI:

1. Center the map view (https://osmdroid.github.io/osmdroid/javadocs/osmdroid-android/release/org/osmdroid/api/IMapController.html#setCenter-org.osmdroid.api.IGeoPoint-)

2. Pass the [MapView](https://osmdroid.github.io/osmdroid/javadocs/osmdroid-android/release/org/osmdroid/views/MapView.html) a [BoundingBox](https://osmdroid.github.io/osmdroid/javadocs/osmdroid-android/release/org/osmdroid/util/BoundingBox.html).
https://osmdroid.github.io/osmdroid/javadocs/osmdroid-android/release/org/osmdroid/views/MapView.html#zoomToBoundingBox-org.osmdroid.util.BoundingBox-boolean-

**Comment by y20k** [comment]
@Hob111 a double tap on the map view would be indeed a "natural" gesture for switching back and forth between the two views.

**Comment by marcoXbresciani** [comment]
I'll try to do my best. Completely newbie on Android side, even though I'm a Java developer.

**Comment by y20k** [comment]
Great. Feel free to contact me, if you have trouble understanding the source code of Trackbook. mail@y20k.org

**Comment by y20k** [comment]
@marcoXbresciani I wired up your BoundingBox. calculation. The app now creates a BoundingBox on save. A long press on the map zooms the track to fit the BoundingBox - the double tap gesture was already taken by the system.

I found two issues so far:
- BoundingBox calculation does not work for a track with only one waypoint
- the BoundingBox would look nicer, if it had some padding around it.

If anyone is interested in testing out the result, here is an APK:
- [trackbook-app-testing-1.2.3-001.apk](https://drive.google.com/open?id=1sLm1hQZf5NdKWyd37PbYfc52pUBW2PSh)

An additional note: The above posted APK will play nice with existing recordings made with an earlier app version, which was installed via Google Play or F-Droid. If you install a version of the app on your phone using Android Studio, older tracks copied back to `/Android/data/org.y20k.trackbook/files/tracks` will lead to a crash.

**Comment by y20k** [comment]
PS. I am going on vacation :palm_tree: soon. An official release of the new feature must wait until mid-september, because I should be around, in case I introduced new bugs.

**Comment by marcoXbresciani** [comment]
> @marcoXbresciani I wired up your BoundingBox. calculation. The app now creates a BoundingBox on save. A long press on the map zooms the track to fit the BoundingBox - the double tap gesture was already taken by the system.
> 
> I found two issues so far:
> 
> * BoundingBox calculation does not work for a track with only one waypoint
> * the BoundingBox would look nicer, if it had some padding around it.
> 
> If anyone is interested in testing out the result, here is an APK:
> 
> * [trackbook-app-testing-1.2.3-001.apk](https://drive.google.com/open?id=1sLm1hQZf5NdKWyd37PbYfc52pUBW2PSh)
> 
> An additional note: The above posted APK will play nice with existing recordings made with an earlier app version, which was installed via Google Play or F-Droid. If you install a version of the app on your phone using Android Studio, older tracks copied back to `/Android/data/org.y20k.trackbook/files/tracks` will lead to a crash.

Cool! :) I don't know how to cope with the single WayPoint... I'll take a look at the code. I might return an home made BoundingBox.
  Ando I also don't know how to add a border. Surely it's nicer than nothing. I'll check about it.

**Comment by marcoXbresciani** [comment]
> PS. I am going on vacation 🌴 soon. An official release of the new feature must wait until mid-september, because I should be around, in case I introduced new bugs.

Lucky you: I've been on holidays ⛩ last month. Already forgot everything. 😭

**Comment by y20k** [comment]
> I don't know how to cope with the single WayPoint

> also don't know how to add a border. 

Yeah. Take your time. Development tends to be slow here. No need to hurry :-)

**Comment by marcoXbresciani** [comment]
> @marcoXbresciani I wired up your BoundingBox. calculation. The app now creates a BoundingBox on save. A long press on the map zooms the track to fit the BoundingBox - the double tap gesture was already taken by the system.
> 
> I found two issues so far:
> 
> * BoundingBox calculation does not work for a track with only one waypoint
> * the BoundingBox would look nicer, if it had some padding around it.
> 
> If anyone is interested in testing out the result, here is an APK:
> 
> * [trackbook-app-testing-1.2.3-001.apk](https://drive.google.com/open?id=1sLm1hQZf5NdKWyd37PbYfc52pUBW2PSh)
> 
> An additional note: The above posted APK will play nice with existing recordings made with an earlier app version, which was installed via Google Play or F-Droid. If you install a version of the app on your phone using Android Studio, older tracks copied back to `/Android/data/org.y20k.trackbook/files/tracks` will lead to a crash.

I've found a fanta-marvellous 🌿  `.increaseByScale` method in the `BoundingBox` itself. Not sure if I should put something like `1.05f` or `.95f` in the scale: doc says `Scale this bounding box by a given factor`. Useful, isn't it? 😀  Also, why I choose 5%? Don't know, seems like a good resize. But maybe there should be something proportional to the actual box size? Let's try with `1.05`...

**Comment by y20k** [comment]
Padding the displayed recording using increaseByScale seems to work fine. The other problem, where a track has very few / to fee waypoints, remains. I will be gone for a couple of weeks. So cannot review any suggestions during that time. Cheers

**Comment by marcoXbresciani** [comment]
> Padding the displayed recording using increaseByScale seems to work fine. The other problem, where a track has very few / too few waypoints, remains. I will be gone for a couple of weeks. So cannot review any suggestions during that time. Cheers

I'll check that. I might think about a few criteria to add a some fake `GeoPoints` to the list in order to have more points to fit.

**Comment by marcoXbresciani** [comment]
Can't continue on this, sorry. I really hope someone can jump in.
